File: Task2/src/data/Animal.cpp
#include <cstdio>
#include <cstring>
#include "Animal.h"
#include "Fish.h"
#include "Bird.h"
#include "Mammal.h"
#include "../utilities/Random.h"

Animal::Animal(const char *name, int mass_gram) : mass_gram_(mass_gram) {
    strcpy(name_, name);
}

const char *Animal::getName() const {
    return name_;
}

int Animal::getMassGram() const {
    return mass_gram_;
}

Animal *Animal::in(FILE *ifst) {
    int k;
    if (fscanf(ifst, "%d", &k) == EOF)
        return nullptr;

    switch (k) {
        case 1:
            return Fish::in(ifst);
        case 2:
            return Bird::in(ifst);
        case 3:
            return Mammal::in(ifst);
        default:
            return nullptr;
    }
}

Animal *Animal::inRnd() {
    int k = Random::RandomRange(1, 3);
    switch (k) {
        case 1:
            return Fish::inRnd();
        case 2:
            return Bird::inRnd();
        case 3:
            return Mammal::inRnd();
        default:
            return nullptr;
    }
}

void Animal::out(FILE *ofst) {
    fprintf(ofst, " Name = %s, mass = %d. Task result: %f\n", name_, mass_gram_, calculateTask());
}

double Animal::calculateTask() const {
    double res = 0;
    for (int i = 0; i < strlen(name_); i++) {
        res += static_cast<int>(name_[i]);
    }

    res /= (double) mass_gram_;

    return res;
}

File: Task3/src/data/animal.py
import random


class Animal:
    def __init__(self):
        self.name = ""
        self.mass_gram = 0

    def __init__(self, name, mass_gram):
        self.name = name
        self.mass_gram = mass_gram

    def get_animal_file(file):
        line_str = file.readline()
        if not line_str:
            return None

        animal_type = int(line_str)

        if (animal_type == 1):
            from data.fish import Fish
            return Fish.get_from_file(file)
        elif (animal_type == 2):
            from data.bird import Bird
            return Bird.get_from_file(file)
        elif (animal_type == 3):
            from data.mammal import Mammal
            return Mammal.get_from_file(file)

    def get_rand_animal():
        animal_type = random.randint(1, 3)
        if (animal_type == 1):
            from data.fish import Fish
            return Fish.get_rand()
        elif (animal_type == 2):
            from data.bird import Bird
            return Bird.get_rand()
        elif (animal_type == 3):
            from data.mammal import Mammal
            return Mammal.get_rand()

    def print(self, file):
        file.write(
            f" Name = {self.name}, mass = {self.mass_gram}. Task result: {self.calculate_task()}\n")

    def calculate_task(self):
        res = 0
        for i in range(len(self.name)):
            res += ord(self.name[i])

        res /= self.mass_gram
        return res
